# Dale Campbell -- Dssa 5001 -- Assignment 1 

## Part 2

### Goals
Find and download a dataset of your choosing, preferably a dataset that piques your interest.  
 
Complete exploratory data analysis on this dataset in a Jupyter notebook. As a reminder, there is 
no set-by-set guide that is applicable to each dataset, as every dataset is different. However, your 
goals in general are to: 
- Understand your variables 
- Clean the dataset 
- Analyze the relationship between variables 
In addition, please meticulously annotate your notebook with your thoughts as you progress 
through the EDA process, writing down new questions that arise as you observe and analyze your 
data as comments. Not only is this best practice, but you’ll be able to reference your analysis in 
detail in the future. The code is secondary to your thought process and insights about your 
findings. 

In [49]:
# Necessary imports
import pandas as pd
import sys

# Opening 'netflix daily top 10.csv'
df = pd.read_csv('../data/netflix daily top 10.csv')

In [2]:
# Now let's view the first few rows to get an idea of the data 
df.head(15)

As of  Rank Year to Date Rank Last Week Rank  \
0   2020-04-01     1                 1              1   
1   2020-04-01     2                 2              -   
2   2020-04-01     3                 3              2   
3   2020-04-01     4                 4              -   
4   2020-04-01     5                 5              4   
5   2020-04-01     6                 6              -   
6   2020-04-01     7                10              -   
7   2020-04-01     8                 7              5   
8   2020-04-01     9                 8              -   
9   2020-04-01    10                 9              -   
10  2020-04-02     1                 1              1   
11  2020-04-02     2                 2              -   
12  2020-04-02     3                 3              2   
13  2020-04-02     4                 -              -   
14  2020-04-02     5                 -              -   

                           Title     Type Netflix Exclusive  \
0   Tiger King: Murder, Mayhem …  TV Show               Yes   
1                          Ozark  TV Show               Yes   
2                   All American  TV Show               NaN   
3                   Blood Father    Movie               NaN   
4                   The Platform    Movie               Yes   
5    Car Masters: Rust to Riches  TV Show               Yes   
6                     Unorthodox  TV Show               Yes   
7                  Love is Blind  TV Show               Yes   
8                        Badland    Movie               NaN   
9                       Uncorked    Movie               Yes   
10  Tiger King: Murder, Mayhem …  TV Show               Yes   
11                         Ozark  TV Show               Yes   
12                  All American  TV Show               NaN   
13                    Nailed It!  TV Show               Yes   
14     How to Fix a Drug Scandal  TV Show               Yes   

   Netflix Release Date  Days In Top 10  Viewership Score  
0          Mar 20, 2020               9                90  
1          Jul 21, 2017               5                45  
2          Mar 28, 2019               9                76  
3          Mar 26, 2020               5                30  
4          Mar 20, 2020               9                55  
5          Sep 14, 2018               4                14  
6          Mar 26, 2020               2                 5  
7          Feb 13, 2020               9                40  
8          Mar 26, 2020               4                11  
9          Mar 27, 2020               4                15  
10         Mar 20, 2020              10               100  
11         Jul 21, 2017               6                54  
12         Mar 28, 2019              10                84  
13          Mar 9, 2018               1                 7  
14          Apr 1, 2020               1                 6

In [3]:
# Honestly, I know they're not exactly the same thing but I prefer to see a truly empty spot as opposed to NaN. Going to replace that and make it easy on my aging eyes. 
# This process won't work for every dataset as I'll need to handle outliers, measure how much data is missing, etc. But for now letsdoit
df = df.fillna("")
df.head(15)

As of  Rank Year to Date Rank Last Week Rank  \
0   2020-04-01     1                 1              1   
1   2020-04-01     2                 2              -   
2   2020-04-01     3                 3              2   
3   2020-04-01     4                 4              -   
4   2020-04-01     5                 5              4   
5   2020-04-01     6                 6              -   
6   2020-04-01     7                10              -   
7   2020-04-01     8                 7              5   
8   2020-04-01     9                 8              -   
9   2020-04-01    10                 9              -   
10  2020-04-02     1                 1              1   
11  2020-04-02     2                 2              -   
12  2020-04-02     3                 3              2   
13  2020-04-02     4                 -              -   
14  2020-04-02     5                 -              -   

                           Title     Type Netflix Exclusive  \
0   Tiger King: Murder, Mayhem …  TV Show               Yes   
1                          Ozark  TV Show               Yes   
2                   All American  TV Show                     
3                   Blood Father    Movie                     
4                   The Platform    Movie               Yes   
5    Car Masters: Rust to Riches  TV Show               Yes   
6                     Unorthodox  TV Show               Yes   
7                  Love is Blind  TV Show               Yes   
8                        Badland    Movie                     
9                       Uncorked    Movie               Yes   
10  Tiger King: Murder, Mayhem …  TV Show               Yes   
11                         Ozark  TV Show               Yes   
12                  All American  TV Show                     
13                    Nailed It!  TV Show               Yes   
14     How to Fix a Drug Scandal  TV Show               Yes   

   Netflix Release Date  Days In Top 10  Viewership Score  
0          Mar 20, 2020               9                90  
1          Jul 21, 2017               5                45  
2          Mar 28, 2019               9                76  
3          Mar 26, 2020               5                30  
4          Mar 20, 2020               9                55  
5          Sep 14, 2018               4                14  
6          Mar 26, 2020               2                 5  
7          Feb 13, 2020               9                40  
8          Mar 26, 2020               4                11  
9          Mar 27, 2020               4                15  
10         Mar 20, 2020              10               100  
11         Jul 21, 2017               6                54  
12         Mar 28, 2019              10                84  
13          Mar 9, 2018               1                 7  
14          Apr 1, 2020               1                 6

In [4]:
# Let's get an idea of what the file looks like by checking the data types
df.dtypes

As of                   object
Rank                     int64
Year to Date Rank       object
Last Week Rank          object
Title                   object
Type                    object
Netflix Exclusive       object
Netflix Release Date    object
Days In Top 10           int64
Viewership Score         int64
dtype: object

Immediately, I think 

    * ['As Of', 'Netflix Release Date'] should be a date types
    * ['Title' , 'Type' , 'Netflix Exclusive'] should be string types
    * ['Year to Date Rank'] should be called YTD Rank

In [5]:
# Let's format these columns
df['As of'] = pd.to_datetime(df['As of'], format='%Y-%m-%d')
df['Netflix Release Date'] = pd.to_datetime(df['Netflix Release Date'], format='%b %d, %Y')
df['YTD Rank'] = df['Year to Date Rank']
df['Title'] = df['Title'].astype("string")
df['Type'] = df['Type'].astype("string")
df['Netflix Exclusive'] = df['Netflix Exclusive'].astype("string")

In [6]:
# Let's show the data again so I don't have to scroll back and forth
df.head(20)

As of  Rank Year to Date Rank Last Week Rank  \
0  2020-04-01     1                 1              1   
1  2020-04-01     2                 2              -   
2  2020-04-01     3                 3              2   
3  2020-04-01     4                 4              -   
4  2020-04-01     5                 5              4   
5  2020-04-01     6                 6              -   
6  2020-04-01     7                10              -   
7  2020-04-01     8                 7              5   
8  2020-04-01     9                 8              -   
9  2020-04-01    10                 9              -   
10 2020-04-02     1                 1              1   
11 2020-04-02     2                 2              -   
12 2020-04-02     3                 3              2   
13 2020-04-02     4                 -              -   
14 2020-04-02     5                 -              -   
15 2020-04-02     6                 4              -   
16 2020-04-02     7                 -              -   
17 2020-04-02     8                 7              -   
18 2020-04-02     9                 5              4   
19 2020-04-02    10                 6              -   

                           Title     Type Netflix Exclusive  \
0   Tiger King: Murder, Mayhem …  TV Show               Yes   
1                          Ozark  TV Show               Yes   
2                   All American  TV Show                     
3                   Blood Father    Movie                     
4                   The Platform    Movie               Yes   
5    Car Masters: Rust to Riches  TV Show               Yes   
6                     Unorthodox  TV Show               Yes   
7                  Love is Blind  TV Show               Yes   
8                        Badland    Movie                     
9                       Uncorked    Movie               Yes   
10  Tiger King: Murder, Mayhem …  TV Show               Yes   
11                         Ozark  TV Show               Yes   
12                  All American  TV Show                     
13                    Nailed It!  TV Show               Yes   
14     How to Fix a Drug Scandal  TV Show               Yes   
15                  Blood Father    Movie                     
16                  The Roommate    Movie                     
17                    Unorthodox  TV Show               Yes   
18                  The Platform    Movie               Yes   
19   Car Masters: Rust to Riches  TV Show               Yes   

   Netflix Release Date  Days In Top 10  Viewership Score YTD Rank  
0            2020-03-20               9                90        1  
1            2017-07-21               5                45        2  
2            2019-03-28               9                76        3  
3            2020-03-26               5                30        4  
4            2020-03-20               9                55        5  
5            2018-09-14               4                14        6  
6            2020-03-26               2                 5       10  
7            2020-02-13               9                40        7  
8            2020-03-26               4                11        8  
9            2020-03-27               4                15        9  
10           2020-03-20              10               100        1  
11           2017-07-21               6                54        2  
12           2019-03-28              10                84        3  
13           2018-03-09               1                 7        -  
14           2020-04-01               1                 6        -  
15           2020-03-26               6                35        4  
16           2020-04-01               1                 4        -  
17           2020-03-26               3                 8        7  
18           2020-03-20              10                57        5  
19           2018-09-14               5                15        6

### Some things I think are worth analyzing here would be:

* Are TV shows or movies more popular at #1 spot. 
    * Are TV shows or movies more popular in general
* What's the average length a TV show stays in the top 10
* What's the average length a movie stays in the top 10
* Average viewership score per ranking of 1-10
* What percent of TV Shows in top 10 are Netflix Exclusives
* What percent of Movies in the top 10 are Netflix Exclusives

I might find other interesting things along the way though. 

In [7]:
# For starters, let's format the df to be ascending by dates, then ranking. This would make it easier to keep track if we have to iterrows later on. 
df = df.sort_values(by=['As of', 'Rank']).reset_index(drop=True)

In [8]:
# Let's see what's more popular at #1 -- TV Show | Movie
df[df['Rank'] == 1].Type.value_counts()

TV Show    494
Movie      216
Name: Type, dtype: Int64

TV Shows are twice as likely as movies to be in the #1 spot. 

In [9]:
df.Type.value_counts()

TV Show            4446
Movie              2611
Stand-Up Comedy      41
Concert/Perf…         2
Name: Type, dtype: Int64

TV Shows are almost twice as likely as movies to appear in the top 10 results. 

In [10]:
# There's not a lot of Stand-Up's appearing in the 10. Let's see what specials we should save for later (fingers crossed I haven't seen them already..)
df[df['Type'] == "Stand-Up Comedy"].Title.value_counts()

Dave Chappelle: The Closer        16
Kevin Hart: Zero F**ks Given      10
Jerry Seinfeld: 23 Hours to…       5
George Lopez: We’ll Do It f…     5
Chris D’Elia: No Pain            3
Bo Burnham: Inside                 2
Name: Title, dtype: Int64

Makes sense we'd see Dave Chappelle and Kevin Hart in there. Legends. 

In [11]:
# Let's agg by individual titles
tempVar = df.groupby(['Title']).agg('sum', numeric_only=True)

# Now let's see the average days in top 10 & viewership scores
print("Average Days In Top 10:\n" + str(round(tempVar['Days In Top 10'].mean(), 2)))
print()
print("Average Viewership Score:\n" + str(round(tempVar['Viewership Score'].mean(), 2)))

Average Days In Top 10:
265.55

Average Viewership Score:
1351.64


So that tells us overall the average # days in the top 10 for movies, TV shows, etc. Let's do the same thing but by Type == Movie & Type == TV Show

In [12]:
tempVar = df[df['Type'] == "Movie"].groupby(['Title']).agg('sum', numeric_only=True)

# Now let's see the average days in top 10 & viewership scores
print("Average Days In Top 10 (Movie):\n" + str(round(tempVar['Days In Top 10'].mean(), 2)))
print()
print("Average Viewership Score (Movie):\n" + str(round(tempVar['Viewership Score'].mean(), 2)))
print()

tempVar = df[df['Type'] == "TV Show"].groupby(['Title']).agg('sum', numeric_only=True)

# Now let's see the average days in top 10 & viewership scores
print("Average Days In Top 10 (TV Show):\n" + str(round(tempVar['Days In Top 10'].mean(), 2)))
print()
print("Average Viewership Score (TV Show):\n" + str(round(tempVar['Viewership Score'].mean(), 2))) 

Average Days In Top 10 (Movie):
44.06

Average Viewership Score (Movie):
286.46

Average Days In Top 10 (TV Show):
549.13

Average Viewership Score (TV Show):
2715.58


From this I we can see TV Shows stay in the top 10 a lot longer. Which naturally makes sense as a TV series has more total watchtime than a movie (statistically speaking)


In [24]:
# Let's find the average viewship score of each spot in the top 10
listVar = list(range(1, 11))

for index in listVar:
    tempVar = df[df['Rank'] == int(index)]
    print("Average Viewership Score of #" + str(index) + " spot:\n" + str(round(tempVar['Viewership Score'].mean(), 2)))

Average Viewership Score of #1 spot:
90.76
Average Viewership Score of #2 spot:
92.23
Average Viewership Score of #3 spot:
93.73
Average Viewership Score of #4 spot:
102.93
Average Viewership Score of #5 spot:
122.22
Average Viewership Score of #6 spot:
147.82
Average Viewership Score of #7 spot:
140.9
Average Viewership Score of #8 spot:
161.86
Average Viewership Score of #9 spot:
147.83
Average Viewership Score of #10 spot:
127.62


In [51]:
%%capture
# Finding the number of movies & TV shows that are Netflix Exclusive vs not
# Capturing output because I don't care about their silly warning messages of deprication. 
tempVar = df[df['Type'] == "Movie"].groupby(['Title']).agg('sum', numeric_only=False)
tempVarTV = df[df['Type'] == "TV Show"].groupby(['Title']).agg('sum', numeric_only=False)


In [52]:
percMovieExc = (tempVar['Netflix Exclusive'].value_counts()[0] / tempVar.shape[0]) * 100
percTVExc = (tempVarTV['Netflix Exclusive'].value_counts()[0] / tempVarTV.shape[0]) * 100

print("Percent of Movies in the top 10 that are Netflix Exclusives:\n" + str(round(percMovieExc, 2)))
print()
print("Percent of TV Shows in the top 10 that are Netflix Exclusives:\n" + str(round(percTVExc, 2)))

Percent of Movies in the top 10 that are Netflix Exclusives:
49.3

Percent of TV Shows in the top 10 that are Netflix Exclusives:
21.55


### Results
* TV Shows are roughly twice as likely to appear in the #1 spot and top 10 overall when compared to Movies. 
* Average time a TV Show & Movie stay in the top 10 

| TV Show     | Movie        |
| ----------- | ----------- |
|  549.13  (days)   | 44.06   (days)     |

* Average Viewership Score / Ranking -- from this we can see the lower the viewership score, the better

| Ranking      | Viewership Score |
| ----------- | ----------- |
| 1      | 90.76       |
| 2   | 92.23        |
| 3   | 93.73        |
| 4   | 102.93        |
| 5   | 122.22        |
| 6   | 147.82        |
| 7   | 140.90       |
| 8   | 161.86        |
| 9   | 147.83       |
| 10   | 127.62        |

* Percent of TV Shows & Movies that are Netflix Exclusives in the top 10

| TV Show     | Movie        |
| ----------- | ----------- |
|  21.55%    | 49.30%      |
